src/roadmap/models.py

```diff
@@ -43,10 +46,42 @@ class System(BaseModel):
     minor: int | None = None
     release_date: date | t.Literal["Unknown"] | None
     retirement_date: date | t.Literal["Unknown"] | None
-    support_status: SupportStatus = SupportStatus.supported
+    support_status: SupportStatus = SupportStatus.unknown
     count: int = 0
     lifecycle_type: LifecycleType
 
+    @model_validator(mode="after")
+    def update_support_status(self):
+        """Validator for setting support status.
+        Expected types/values of start_date and end_date are:
+            - str(Unknown)
+            - None
+            - date(YYYY-MM-DD)
+        """
+        today = date.today()
+
+        if self.release_date != "Unknown" and self.release_date is not None:
+            if self.release_date > today:
+                self.support_status = SupportStatus.upcoming
+                return self
+
+        if self.retirement_date != "Unknown" and self.retirement_date is not None:
+            if self.retirement_date < today:
+                self.support_status = SupportStatus.retired
+                return self
+
+            six_months_date = self.retirement_date - timedelta(days=180)
+            if six_months_date <= today:
+                self.support_status = SupportStatus.six_months
+            else:
+                self.support_status = SupportStatus.supported
+
+        else:
+            self.support_status = SupportStatus.unknown
+            return self
+
+        return self
```

Comment:
Here's a simplified approach. 
```python
self.support_status = SupportStatus.unknown # This may be unnecessary but prevents return the passed in value.
if isinstance(self.start_date, date):
    if self.start_date > self.current_date:
        self.support_status = SupportStatus.upcoming
        return self

if isinstance(self.end_date, date):
    if self.end_date < self.current_date:
        self.support_status = SupportStatus.retired
        return self

    six_months_date = self.end_date - timedelta(days=180)
    if six_months_date <= self.current_date:
        self.support_status = SupportStatus.six_months
        return self

    self.support_status = SupportStatus.supported

return self
```

Reply:
If you wanted to pull this code out to use in both classes, you could turn it into a function and have this method call the function. It's not a big deal since it's only in two places, but that's an option.
```python
    @model_validator(mode="after")
    def update_support_status(self):
        return _validator_support_status(self)
```

Reply:
That's exactly what I was thinking when I started applying the suggestions. "Why I haven't did this as a function and I need to change it twice?!"

I'm moving the status calculation into the `common.py`

Thanks for the review! :)

Reply:
> So not in common.py there is an circular import...

We may need to rethink the organization a bit to avoid these issues. If we're getting enough utility functions, we could move those to a different module. Or avoid putting models in `common`.

> And instead of the isinstance() I'm using the not in (None, SupportStatus.unknown) expression. 

I switched to `isinstance` to make the type checker happy. I like the later expression better.